### ISSUE TITLE
Sprout ABI conflicts hard to debug without code reading

### ISSUE THREAD

**Issue by chris-elford-metaswitch** [Metaswitch/sprout#1485]
Saw a problem on the mailing list where a user had different versions of Sprout and the Memento Sproutlet plugin. They were built against different copies of sproutlet_options.h. Sprout set the enabled_memento flag to True, but Memento looked in the wrong place and concluded that it was false.

I was only able to spot this after code reading, and deducing that the only way (in this particular sproutlet) that the logged code path could be hit was due to this mismatch.

It's possible to spot this (if you know what you are looking for) by running `dpkg -l | grep -e memento-as -e "\ssprout[-\s]"` and comparing versions, but there is nothing in the troubleshooting guide to suggest doing that.

One solution is to introduce ABI versioning for the Sprout/Sproutlet interface, and to refuse to load sproutlets (writing a noisy log) built to the wrong ABI.

Users will get themselves into this situation when they install Sprout, and install a Sproutlet later after running `sudo apt-get update`. Running a [Clearwater upgrade](http://clearwater.readthedocs.io/en/stable/Upgrading_a_Clearwater_deployment.html) fixed the problem. This could be better documented, but I still think that diagnostics would need to be improved.

**Comment by graemerobertson** [comment]
We do have API versioning - see https://github.com/Metaswitch/sprout/blob/99cdf08d5bdbcbc8a1a4d313c6861e0442d524da/include/sproutlet.h#L56 and https://github.com/Metaswitch/sprout/blob/386ecaf14f569a8444e660f14090dc8f41c9166c/src/pluginloader.cpp#L122. Unfortunately we've never used it (we're still at version 1 of the Sproutlet API!).

This doesn't completely fix this issue since I guess the ABI can break in ways other than the underlying API changing, but I'm guessing it would fix the most likely cause of us hitting this issue, so we can probably reduce to low priority.

**Comment by chris-elford-metaswitch** [comment]
That helps, but to fix the issue I saw, we would need to change API version whenever we change the options in sproutlet_options.h.

**Comment by graemerobertson** [comment]
Agreed, but I think we should be doing that seeing as they're part of the API.

**Comment by graemerobertson** [comment]
We should write up our policy on changing the API version and check that the API version checking actually works as desired before lowering the priority of this issue.

**Comment by chris-elford-metaswitch** [comment]
Agreed. If we put the right process/docs in place this becomes much less urgent.
